cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerModificationService.java

```diff
@@ -473,22 +519,76 @@ private void activateParcels(ClustersResourceApi clustersResourceApi) throws Api
         LOGGER.debug("Parcels are activated on upscaled hosts.");
     }
 
-    private ApiHostRefList createUpscaledHostRefList(List<String> upscaleHostNames, List<ApiHost> hosts) {
-        LOGGER.debug("Creating ApiHostRefList from upscaled hosts.");
+    private ApiHostRefList createUpscaledHostRefList(String hostGroupName, Map<String, InstanceMetaData> upscaleInstancesMap,
+            Map<String, ApiHost> upscaleHostsMap) {
+        LOGGER.debug("Creating ApiHostRefList from upscaled hosts. Host group: [{}].", hostGroupName);
         ApiHostRefList body = new ApiHostRefList();
-        upscaleHostNames.forEach(hostname -> {
-            hosts.stream()
-                    .filter(host -> hostname.equalsIgnoreCase(host.getHostname()))
-                    .findFirst().ifPresent(apiHost -> {
-                String hostId = apiHost.getHostId();
-                body.addItemsItem(
-                        new ApiHostRef().hostname(hostname).hostId(hostId));
-            });
-        });
-        LOGGER.debug("Created ApiHostRefList from upscaled hosts. Host count: [{}]", body.getItems().size());
+        upscaleHostsMap.forEach((hostname, host) ->
+                Optional.ofNullable(upscaleInstancesMap.get(hostname))
+                        .ifPresent(instance -> {
+                            ApiHostRef apiHostRef = new ApiHostRef()
+                                    .hostname(instance.getDiscoveryFQDN())
+                                    .hostId(host.getHostId());
+                            body.addItemsItem(apiHostRef);
+                        }));
+        LOGGER.debug("Created ApiHostRefList from upscaled hosts. Host group: [{}]. Host count: [{}]", hostGroupName, body.getItems().size());
         return body;
     }
 
+    private void setHostRackIdForOutdatedClusterHosts(String hostGroupName, Collection<InstanceMetaData> instanceMetaDatas, Set<String> clusterHostnames,
+            List<ApiHost> hosts) throws ApiException {
+        Map<String, InstanceMetaData> clusterInstancesMap = getInstancesMap(hostGroupName, clusterHostnames, instanceMetaDatas, false);
+        Map<String, ApiHost> clusterHostsMap = getHostsMap(hostGroupName, clusterInstancesMap, hosts, "cluster");
+        setHostRackIdBatch(hostGroupName, clusterInstancesMap, clusterHostsMap, "outdated cluster");
+    }
+
+    private void setHostRackIdBatch(String hostGroupName, Map<String, InstanceMetaData> instancesMap, Map<String, ApiHost> hostsMap, String hostType)
```

Comment:
I agree, it looks odd. On the other hand, I fear it would be complicated to match the lines belonging to the same upscale flow. The hostgroup being processed in not part of the MDC. I am open to suggestions if you have a better idea how to handle this.

Reply:
I meant you should put a log line into the caller method, where you pass this variable. A log line there, and you leave the current one without this parameter and you should be able to make the connection base on traceid/thread/whatever when investigating.

Reply:
Okay, the request ID / span ID seems to be enough to correlate these logs. Will refactor this accordingly.